<h1 style="color:#C39BD3; border-bottom: 3px solid #C39BD3; padding-bottom: 8px;">
AI & ML Course with BinX — Day 1 — Descriptive Statistics
</h1>

## <span style="color:#7FB3D5">1.1 Why Statistics Comes Before Modeling</span>

<blockquote style="border-left: 3px solid #7FB3D5; padding-left: 12px; margin-left: 0;">
A machine learning model is, at its core, a summary of patterns in data.
If you cannot describe a dataset statistically — its center, its spread,
its shape — you cannot judge whether a model's behavior is reasonable.
Descriptive statistics is the vocabulary for talking about data precisely,
and it underpins every EDA and evaluation step that follows.
</blockquote>

## <span style="color:#7FB3D5">1.2 Measures of Central Tendency</span>

<table style="border-collapse: collapse; width:100%;">
<tr>
<th style="border: 1px solid #58D68D; padding:8px;">Measure</th>
<th style="border: 1px solid #58D68D; padding:8px;">What It Is</th>
<th style="border: 1px solid #58D68D; padding:8px;">When to Use It</th>
</tr>
<tr>
<td style="border: 1px solid #58D68D; padding:8px;">Mean</td>
<td style="border: 1px solid #58D68D; padding:8px;">The arithmetic average</td>
<td style="border: 1px solid #58D68D; padding:8px;">Symmetric data with no extreme outliers</td>
</tr>
<tr>
<td style="border: 1px solid #58D68D; padding:8px;">Median</td>
<td style="border: 1px solid #58D68D; padding:8px;">The middle value when sorted</td>
<td style="border: 1px solid #58D68D; padding:8px;">Skewed data or data with outliers (more robust)</td>
</tr>
<tr>
<td style="border: 1px solid #58D68D; padding:8px;">Mode</td>
<td style="border: 1px solid #58D68D; padding:8px;">The most frequent value</td>
<td style="border: 1px solid #58D68D; padding:8px;">Categorical data, or finding the most common value</td>
</tr>
</table>

**Key insight:** the mean is pulled toward outliers, while the median is not.

In [6]:
import numpy as np

data = np.array([10, 12, 12, 13, 100])
print("\033[92mMean:\033[0m", np.mean(data))
print("\033[94mMedian:\033[0m", np.median(data))

Mean: 29.4
Median: 12.0


## <span style="color:#7FB3D5">1.3 Measures of Spread</span>

<table style="border-collapse: collapse; width:100%;">
<tr>
<th style="border: 1px solid #58D68D; padding:8px;">Measure</th>
<th style="border: 1px solid #58D68D; padding:8px;">Meaning</th>
</tr>
<tr>
<td style="border: 1px solid #58D68D; padding:8px;">Range</td>
<td style="border: 1px solid #58D68D; padding:8px;">Maximum minus minimum — simple but very sensitive to outliers</td>
</tr>
<tr>
<td style="border: 1px solid #58D68D; padding:8px;">Variance</td>
<td style="border: 1px solid #58D68D; padding:8px;">The average squared distance from the mean</td>
</tr>
<tr>
<td style="border: 1px solid #58D68D; padding:8px;">Standard deviation</td>
<td style="border: 1px solid #58D68D; padding:8px;">The square root of variance — spread in the same units as the data</td>
</tr>
<tr>
<td style="border: 1px solid #58D68D; padding:8px;">IQR</td>
<td style="border: 1px solid #58D68D; padding:8px;">The range of the middle 50% of the data (Q3 - Q1), robust to outliers</td>
</tr>
</table>

In [7]:
data = np.array([10, 12, 12, 13, 100])

print("\033[92mStandard deviation:\033[0m", np.std(data))
print("\033[94mVariance:\033[0m", np.var(data))

q1, q3 = np.percentile(data, [25, 75])
print("\033[93mIQR:\033[0m", q3 - q1)

Standard deviation: 35.31345352694918
Variance: 1247.0399999999997
IQR: 1.0


## <span style="color:#7FB3D5">1.4 Percentiles and Quartiles</span>

<blockquote style="border-left: 3px solid #7FB3D5; padding-left: 12px; margin-left: 0;">
A percentile marks the value below which a given percentage of the data
falls — the 25th percentile (Q1), 50th (the median), and 75th (Q3) split
the data into quarters. These are the exact values a box plot visualizes.
</blockquote>

In [8]:
data = np.array([10, 12, 12, 13, 100])

q1 = np.percentile(data, 25)
q2 = np.percentile(data, 50)
q3 = np.percentile(data, 75)

print("\033[92mQ1 (25th percentile):\033[0m", q1)
print("\033[94mQ2 / Median (50th percentile):\033[0m", q2)
print("\033[93mQ3 (75th percentile):\033[0m", q3)

Q1 (25th percentile): 12.0
Q2 / Median (50th percentile): 12.0
Q3 (75th percentile): 13.0


## <span style="color:#F1948A">🧪 Hands-On Lab: Describing a Real Dataset</span>

### <span style="color:#F7DC6F">Step 1: Load a numeric column from the dataset</span>

In [9]:
import pandas as pd

lab_df = pd.read_csv("data.csv")
co2_values = lab_df["value"]

print("\033[96mNumber of values:\033[0m", co2_values.shape[0])
co2_values.head()

Number of values: 13953


0    11092.675
1    11576.719
2    12713.489
3    12178.107
4    11840.743
Name: value, dtype: float64

### <span style="color:#F7DC6F">Step 2: Compute mean, median, mode, standard deviation, and IQR</span>

In [10]:
mean_val = co2_values.mean()
median_val = co2_values.median()
mode_val = co2_values.mode()[0]
std_val = co2_values.std()

q1 = np.percentile(co2_values, 25)
q3 = np.percentile(co2_values, 75)
iqr_val = q3 - q1

print("\033[92mMean:\033[0m", mean_val)
print("\033[94mMedian:\033[0m", median_val)
print("\033[93mMode:\033[0m", mode_val)
print("\033[95mStandard deviation:\033[0m", std_val)
print("\033[96mIQR:\033[0m", iqr_val)

Mean: 825498.3433543197
Median: 13900.0
Mode: 47.671
Standard deviation: 2788923.2126782313
IQR: 163177.933


### <span style="color:#F7DC6F">Step 3: Which measure better represents a "typical" value?</span>

<blockquote style="border-left: 3px solid #F7DC6F; padding-left: 12px; margin-left: 0;">
<b>Justification:</b> The median (13,900) is a far better representation of a
"typical" CO2 value than the mean (825,498). The dataset mixes individual
countries with large aggregated regions (e.g. "World", "High income",
"OECD members"), which have extremely large CO2 values. These act as
outliers that pull the mean far above what most individual records
actually look like. The median, being robust to outliers, reflects the
center of the bulk of the data far more accurately.
</blockquote>

### <span style="color:#F7DC6F">Step 4: Summary of center and spread</span>

<blockquote style="border-left: 3px solid #F7DC6F; padding-left: 12px; margin-left: 0;">
<b>Summary:</b> The CO2 emission values in this dataset are heavily
right-skewed. While the average value sits around 825,498, the typical
(median) value is only about 13,900 — meaning most records represent much
smaller emitters than the mean suggests. The standard deviation
(~2,788,923) is enormous relative to the median, confirming a wide spread
driven by a small number of extremely high values. The IQR (~163,178) is
far smaller than the standard deviation, showing that the middle 50% of
the data is much more tightly clustered than the full range implies. In
short: most records are small, but a handful of very large aggregated
regions dominate the overall statistics.
</blockquote>